File: src/table/table.go
package table

type CostTable struct {
	UpgradesCost struct {
		Damage struct {
			Value        []int  `json:"value"`
			ResourceType string `json:"resourceType"`
		} `json:"damage"`
		Attack struct {
			Value        []int  `json:"value"`
			ResourceType string `json:"resourceType"`
		} `json:"attack"`
		Health struct {
			Value        []int  `json:"value"`
			ResourceType string `json:"resourceType"`
		} `json:"health"`
		Defence struct {
			Value        []int  `json:"value"`
			ResourceType string `json:"resourceType"`
		} `json:"defence"`
	} `json:"upgradesCost"`
	UpgradeMultiplier struct {
		Damage  []float64 `json:"damage"`
		Attack  []float64 `json:"attack"`
		Health  []float64 `json:"health"`
		Defence []float64 `json:"defence"`
	} `json:"upgradeMultiplier"`
	UpgradeBaseStats struct {
		Damage struct {
			Min []int `json:"min"`
			Max []int `json:"max"`
		} `json:"damage"`
		Attack  []float64 `json:"attack"`
		Health  []int     `json:"health"`
		Defence []float64 `json:"defence"`
	} `json:"upgradeBaseStats"`
	Bonuses struct {
		StrongerMinions struct {
			MaxLevel     int       `json:"maxLevel"`
			Cost         []int     `json:"cost"`
			ResourceType string    `json:"resourceType"`
			Value        []float64 `json:"value"`
		} `json:"strongerMinions"`
		TowerMultiAttacks struct {
			MaxLevel     int    `json:"maxLevel"`
			Cost         []int  `json:"cost"`
			ResourceType string `json:"resourceType"`
			Value        []int  `json:"value"`
		} `json:"towerMultiAttacks"`
		TowerAura struct {
			MaxLevel     int           `json:"maxLevel"`
			Cost         []int         `json:"cost"`
			ResourceType string        `json:"resourceType"`
			Value        []interface{} `json:"value"`
		} `json:"towerAura"`
		TowerVolley struct {
			MaxLevel     int           `json:"maxLevel"`
			Cost         []int         `json:"cost"`
			ResourceType string        `json:"resourceType"`
			Value        []interface{} `json:"value"`
		} `json:"towerVolley"`
		XpSeeking struct {
			MaxLevel     int    `json:"maxLevel"`
			Cost         []int  `json:"cost"`
			ResourceType string `json:"resourceType"`
			Value        []int  `json:"value"`
		} `json:"xpSeeking"`
		TomeSeeking struct {
			MaxLevel     int       `json:"maxLevel"`
			Cost         []int     `json:"cost"`
			ResourceType string    `json:"resourceType"`
			Value        []float64 `json:"value"`
		} `json:"tomeSeeking"`
		EmeraldsSeeking struct {
			MaxLevel     int       `json:"maxLevel"`
			Cost         []int     `json:"cost"`
			ResourceType string    `json:"resourceType"`
			Value        []float64 `json:"value"`
		} `json:"emeraldsSeeking"`
		LargerResourceStorage struct {
			MaxLevel     int    `json:"maxLevel"`
			Cost         []int  `json:"cost"`
			ResourceType string `json:"resourceType"`
			Value        []int  `json:"value"`
		} `json:"largerResourceStorage"`
		LargerEmeraldsStorage struct {
			MaxLevel     int    `json:"maxLevel"`
			Cost         []int  `json:"cost"`
			ResourceType string `json:"resourceType"`
			Value        []int  `json:"value"`
		} `json:"largerEmeraldsStorage"`
		EfficientResource struct {
			MaxLevel     int       `json:"maxLevel"`
			Cost         []int     `json:"cost"`
			ResourceType string    `json:"resourceType"`
			Value        []float64 `json:"value"`
		} `json:"efficientResource"`
		EfficientEmeralds struct {
			MaxLevel     int       `json:"maxLevel"`
			Cost         []int     `json:"cost"`
			ResourceType string    `json:"resourceType"`
			Value        []float64 `json:"value"`
		} `json:"efficientEmeralds"`
		ResourceRate struct {
			MaxLevel     int    `json:"maxLevel"`
			Cost         []int  `json:"cost"`
			ResourceType string `json:"resourceType"`
			Value        []int  `json:"value"`
		} `json:"resourceRate"`
		EmeraldsRate struct {
			MaxLevel     int    `json:"maxLevel"`
			Cost         []int  `json:"cost"`
			ResourceType string `json:"resourceType"`
			Value        []int  `json:"value"`
		} `json:"emeraldsRate"`
	} `json:"bonuses"`
}

type TerritoryProperty struct {
	Upgrades     TerritoryPropertyUpgradeData
	Bonuses      TerritoryPropertyBonusesData
	Tax          Tax
	Border       string
	TradingStyle string
	HQ           bool
}

type TerritoryPropertyUpgradeData struct {
	Damage  int
	Attack  int
	Health  int
	Defence int
}

type TerritoryPropertyBonusesData struct {
	StrongerMinions       int
	TowerMultiAttack      int
	TowerAura             int
	TowerVolley           int
	LargerResourceStorage int
	LargerEmeraldStorage  int
	EfficientResource     int
	EfficientEmerald      int
	ResourceRate          int
	EmeraldRate           int
}

type Tax struct {
	Ally   int
	Others int
}

type TerritoryResource struct {
	Emerald int
	Ore     int
	Wood    int
	Fish    int
	Crop    int
}

type Territory struct {
	Name                   string
	BaseResourceProduction TerritoryResource
	Property               TerritoryProperty
	Storage                TerritoryResourceStorage
	TransversingResource   TerritoryResource
	ResourceProduction     TerritoryResource
	TerritoryUsage         TerritoryResource
	Type                   string
	TradingRoutes          []string
	RouteToHQ              []string
	Claim                  bool
	ID                     int
}

type TerritoryResourceStorage struct {
	Capacity struct {
		Emerald int
		Ore     int
		Wood    int
		Fish    int
		Crop    int
	}
	Current struct {
		Emerald int
		Ore     int
		Wood    int
		Fish    int
		Crop    int
	}
}

type RawTerritoryData struct {
	Resources struct {
		Emeralds int `json:"emeralds"`
		Ore      int `json:"ore"`
		Crops    int `json:"crops"`
		Fish     int `json:"fish"`
		Wood     int `json:"wood"`
	} `json:"resources"`
	TradingRoutes []string `json:"Trading Routes"`
	Type          string
}

type TerritoryUpdateData struct {
	Property TerritoryProperty
}

func (t *Territory) Set(d TerritoryUpdateData) *Territory {
	// validate the territory before setting
	// Damage, Attack, Health and Defence must be between 0 and 11
	// StrongerMinions 0 - 3, Tower MultiAttack 0 - 1, Tower Aura and Volley 0 - 3
	// Larger Emerald and Resource storage 0 - 6, Efficient Resource 0 - 6, Efficient Emerald 0 - 3 and Resource and Emerald Rate 0 - 3
	if d.Property.Upgrades.Damage < 0 || d.Property.Upgrades.Damage > 11 {
		return t
	} else if d.Property.Upgrades.Attack < 0 || d.Property.Upgrades.Attack > 11 {
		return t
	} else if d.Property.Upgrades.Health < 0 || d.Property.Upgrades.Health > 11 {
		return t
	} else if d.Property.Upgrades.Defence < 0 || d.Property.Upgrades.Defence > 11 {
		return t
	} else if d.Property.Bonuses.StrongerMinions < 0 || d.Property.Bonuses.StrongerMinions > 3 {
		return t
	} else if d.Property.Bonuses.TowerMultiAttack < 0 || d.Property.Bonuses.TowerMultiAttack > 1 {
		return t
	} else if d.Property.Bonuses.TowerAura < 0 || d.Property.Bonuses.TowerAura > 3 {
		return t
	} else if d.Property.Bonuses.TowerVolley < 0 || d.Property.Bonuses.TowerVolley > 3 {
		return t
	} else if d.Property.Bonuses.LargerResourceStorage < 0 || d.Property.Bonuses.LargerResourceStorage > 6 {
		return t
	} else if d.Property.Bonuses.LargerEmeraldStorage < 0 || d.Property.Bonuses.LargerEmeraldStorage > 6 {
		return t
	} else if d.Property.Bonuses.EfficientResource < 0 || d.Property.Bonuses.EfficientResource > 6 {
		return t
	} else if d.Property.Bonuses.EfficientEmerald < 0 || d.Property.Bonuses.EfficientEmerald > 3 {
		return t
	} else if d.Property.Bonuses.ResourceRate < 0 || d.Property.Bonuses.ResourceRate > 3 {
		return t
	} else if d.Property.Bonuses.EmeraldRate < 0 || d.Property.Bonuses.EmeraldRate > 3 {
		return t
	}

	t.Property = d.Property
	return t
}

func (t *Territory) Update() *Territory {
	// update the territory stats and cost

	return t
}

func (t *Territory) CloseBorder() *Territory {
	t.Property.Border = "Closed"
	return t
}

func (t *Territory) OpenBorder() *Territory {
	t.Property.Border = "Open"
	return t
}

func (t *Territory) Fastest() *Territory {
	t.Property.TradingStyle = "Fastest"
	return t
}

func (t *Territory) Cheapest() *Territory {
	t.Property.TradingStyle = "Cheapest"
	return t
}

func (t *Territory) ToggleBorder() *Territory {
	if t.Property.Border == "Closed" {
		t.Property.Border = "Open"
	} else {
		t.Property.Border = "Closed"
	}
	return t
}

func (t *Territory) UnsetHQ() *Territory {
	t.Property.HQ = false
	return t
}

func (t *Territory) SetAllyTax(n int) *Territory {
	// tax has to be within 5 and 60
	if n < 5 || n > 60 {
		return t
	}
	t.Property.Tax.Ally = n
	return t
}

func (t *Territory) SetOthersTax(n int) *Territory {
	// tax has to be within 5 and 60
	if n < 5 || n > 60 {
		return t
	}
	t.Property.Tax.Others = n
	return t
}

func (t *Territory) AddTradingRoute(r string) *Territory {
	t.TradingRoutes = append(t.TradingRoutes, r)
	return t
}

func (t *Territory) SetHQ() *Territory {
	t.Property.HQ = true
	return t
}

File: src/table/table.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize)]
pub struct CostTable {
    upgrades_cost: UpgradesCost,
    upgrade_multiplier: UpgradeMultiplier,
    upgrade_base_stats: UpgradeBaseStats,
    bonuses: Bonuses,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct UpgradesCost {
    damage: UpgradeCost,
    attack: UpgradeCost,
    health: UpgradeCost,
    defence: UpgradeCost,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct UpgradeCost {
    value: Vec<i32>,
    resource_type: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct UpgradeMultiplier {
    damage: Vec<f64>,
    attack: Vec<f64>,
    health: Vec<f64>,
    defence: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct UpgradeBaseStats {
    damage: BaseStats,
    attack: Vec<f64>,
    health: Vec<i32>,
    defence: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct BaseStats {
    min: Vec<i32>,
    max: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Bonuses {
    stronger_minions: StrongerMinions,
    tower_multi_attacks: TowerMultiAttacks,
    tower_aura: TowerAura,
    tower_volley: TowerVolley,
    xp_seeking: XpSeeking,
    tome_seeking: TomeSeeking,
    emeralds_seeking: EmeraldsSeeking,
    larger_resource_storage: LargerResourceStorage,
    larger_emeralds_storage: LargerEmeraldsStorage,
    efficient_resource: EfficientResource,
    efficient_emeralds: EfficientEmeralds,
    resource_rate: ResourceRate,
    emeralds_rate: EmeraldsRate,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct StrongerMinions {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TowerMultiAttacks {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TowerAura {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TowerVolley {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct XpSeeking {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TomeSeeking {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct EmeraldsSeeking {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct LargerResourceStorage {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct LargerEmeraldsStorage {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct LargerEmeraldsStorage {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct EfficientResource {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct EfficientEmeralds {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<f64>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ResourceRate {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct EmeraldsRate {
    max_level: i32,
    cost: Vec<i32>,
    resource_type: String,
    value: Vec<i32>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TerritoryProperty {
  upgrade: TerritoryPropertyUpgradeData,
  bonuses: TerritoryPropertyBonusesData,
  tax: i32,
  border: String,
  trading_style: String,
  hq: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TerritoryPropertyUpgradeData {
  damage: i32,
  attack: i32,
  health: i32,
  defence: i32,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct TerritoryPropertyBonusesData {
  stronger_minions: i32,
  tower_multi_attack: i32,
  tower_aura: i32,
  tower_volley: i32,
  larger_resource_storage: i32,
  larger_emeralds_storage: i32,
  efficient_resource: i32,
  efficient_emeralds: i32,
  resource_rate: i32,
  emeralds_rate: i32,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Tax {
  ally: i32,
  other: i32
}

/*
type TerritoryResource struct {
	Emerald int
	Ore     int
	Wood    int
	Fish    int
	Crop    int
}

*/
#[derive(Debug, Deserialize, Serialize)]
pub struct TerritoryResource {
  emeralds: i32,
  ore: i32,
  wood: i32,
  fish: i32,
  crops: i32,
}

/*
type Territory struct {
	Name                   string
	BaseResourceProduction TerritoryResource
	Property               TerritoryProperty
	Storage                TerritoryResourceStorage
	ResourceProduction     TerritoryResource
	TerritoryUsage         TerritoryResource
	Type                   string
	TradingRoutes          []string
	RouteToHQ              []string
}
*/

#[derive(Debug, Deserialize, Serialize)]
pub struct Territory {
  name: String,
  base_resource_production: TerritoryResource,
  property: TerritoryProperty,
  storage: TerritoryResourceStorage,
  resource_production: TerritoryResource,
  territory_usage: TerritoryResource,
  territory_type: String,
  trading_routes: Vec<String>,
  route_to_hq: Vec<String>,
}

/*
type TerritoryResourceStorage struct {
	Capacity struct {
		Emerald int
		Ore     int
		Wood    int
		Fish    int
		Crop    int
	}
	Current struct {
		Emerald int
		Ore     int
		Wood    int
		Fish    int
		Crop    int
	}
}
*/

#[derive(Debug, Deserialize, Serialize)]
pub struct TerritoryResourceStorage {
  capacity: TerritoryResource,
  current: TerritoryResource,
}

/*
type RawTerritoryData struct {
	Resources struct {
		Emeralds int `json:"emeralds"`
		Ore      int `json:"ore"`
		Crops    int `json:"crops"`
		Fish     int `json:"fish"`
		Wood     int `json:"wood"`
	} `json:"resources"`
	TradingRoutes []string `json:"Trading Routes"`
	Type          string
}
*/

#[derive(Debug, Deserialize, Serialize)]
pub struct RawTerritoryData {
  resources: TerritoryResource,
  trading_routes: Vec<String>,
  territory_type: String,
}
/*
type TerritoryUpdateData struct {
	Territory string
	Property  TerritoryProperty
}
*/

#[derive(Debug, Deserialize, Serialize)]
pub struct TerritoryUpdateData {
  territory: String,
  property: TerritoryProperty,
}

fn main() {
    // Deserialize the JSON into a CostTable struct
    
    let cost_table: CostTable = serde_json::from_str(json_data).unwrap();

    // Print the deserialized cost_table struct
    println!("{:#?}", cost_table);
}